src/analysis/processing/qgsalgorithmclimb.cpp

```diff
@@ -0,0 +1,204 @@
+}
+
+QString QgsClimbAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "Calculates the total climb and descent along line geometries.\n\n"
+                      "Input layer must have Z values present. If Z values are not available, the \"Drape\" (set Z "
+                      "value from raster) algorithm may be used to add Z values from a DEM layer.\n\n"
+                      "The output layer is a copy of the input layer with additional fields that contain the total "
+                      "climb, total descent, the minimum elevation and the maximum elevation for each line geometry."
+                      "If the input layer contains fields with the same names as these added fields, they will be "
+                      "renamed (field names will be altered to \"name_2\", \"name_3\", etc, finding the first "
+                      "non-duplicate name)." );
+}
+
+QgsClimbAlgorithm *QgsClimbAlgorithm::createInstance() const
+{
+  return new QgsClimbAlgorithm();
+}
+
+void QgsClimbAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ), QObject::tr( "Line layer" ), QList<int>() << static_cast<int>( Qgis::ProcessingSourceType::VectorLine ) ) );
+  addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "OUTPUT" ), QObject::tr( "Climb layer" ) ) );
+  // TODO QGIS 4.0 harmonize output names with the rest of algorithms (use underscore to separate words)
+  addOutput( new QgsProcessingOutputNumber( QStringLiteral( "TOTALCLIMB" ), QObject::tr( "Total climb" ) ) );
+  addOutput( new QgsProcessingOutputNumber( QStringLiteral( "TOTALDESCENT" ), QObject::tr( "Total descent" ) ) );
+  addOutput( new QgsProcessingOutputNumber( QStringLiteral( "MINELEVATION" ), QObject::tr( "Minimum elevation" ) ) );
+  addOutput( new QgsProcessingOutputNumber( QStringLiteral( "MAXELEVATION" ), QObject::tr( "Maximum elevation" ) ) );
+}
+
+QVariantMap QgsClimbAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  std::unique_ptr<QgsProcessingFeatureSource> source( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+  if ( !source )
+  {
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "INPUT" ) ) );
+  }
+
+  if ( !QgsWkbTypes::hasZ( source->wkbType() ) )
+  {
+    throw QgsProcessingException( QObject::tr( "The layer does not have Z values. If you have a DEM, use the Drape algorithm to extract Z values." ) );
+  }
+
+  QgsFields outputFields = source->fields();
+  QgsFields newFields;
+  newFields.append( QgsField( QStringLiteral( "climb" ), QMetaType::Type::Double ) );
+  newFields.append( QgsField( QStringLiteral( "descent" ), QMetaType::Type::Double ) );
+  newFields.append( QgsField( QStringLiteral( "min_elev" ), QMetaType::Type::Double ) );
+  newFields.append( QgsField( QStringLiteral( "max_elev" ), QMetaType::Type::Double ) );
```

Comment:
better not change these -- it might break existing workflows:

```suggestion
  newFields.append( QgsField( QStringLiteral( "minelev" ), QMetaType::Type::Double ) );
  newFields.append( QgsField( QStringLiteral( "maxelev" ), QMetaType::Type::Double ) );
```